Support encoder-only models without KV-Cache

maxdebayser opened the pull request:
Add support for encoder models such as BERT which don't support a KV cache due to the non-causal attention. Since the KV Cache Spec is used to build the attention metadata for decoder models, this PR initializes the attention metadata builds for encoder-only models directly from the layers and adds a function to build the attention metadata.

This PR combines elements of PRs
https://github.com/vllm-project/vllm/pull/21088
and https://github.com/vllm-project/vllm/pull/19988

Summary of changes:

**Flash Attention Backend:**
- Implement encoder self-attention support without using KV cache

**Scheduler:**
- Disable chunked prefill for models without KV cache

**GPU Model Runner:**
- Implement encoder-only attention metadata building for self-attention

Related to:
- V0 deprecation: #18571
- 2025 Q3 roadmap: #20336

This PR is co-authored with @russellb. It borrows all of the encoder-only attention code from his PR https://github.com/vllm-project/vllm/pull/21088 but leaves out the cross-encoder and encoder attention.

cc: @DarkLight1337 

FIX #18052

maxdebayser committed:
Support encoder-only models without KV-Cache

Add support for encoder models such as BERT which don't support
a KV cache due to the non-causal attention. Since the KV Cache
Spec is used to build the attention metadata for decoder models,
this PR initializes the attention metadata builds for encoder-only
models directly from the layers and adds a function to build the
attention metadata.

This PR combines elements of PRs
https://github.com/vllm-project/vllm/pull/21088
and https://github.com/vllm-project/vllm/pull/19988

Summary of changes:

**Flash Attention Backend:**
- Implement encoder self-attention support without using KV cache

**Scheduler:**
- Disable chunked prefill for models without KV cache

**GPU Model Runner:**
- Implement encoder-only attention metadata building for self-attention

Related to:
- V0 deprecation: #18571
- 2025 Q3 roadmap: #20336

Signed-off-by: Max de Bayser <maxdebayser@gmail.com>
Co-authored-by: Russell Bryant <rbryant@redhat.com>

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
@DarkLight1337 this PR should enable support all bert models except for the classifier models that require token type ids. But that can be left as a future PR as there are several implementation alternatives. Since the KV cache is disable in this PR, it require much less changes than PR https://github.com/vllm-project/vllm/pull/19988

DarkLight1337 commented on tests/entrypoints/openai/test_rerank.py:
```diff
@@ -124,4 +124,4 @@ def test_invocations(server: RemoteOpenAIServer):
         assert rerank_result.keys() == invocations_result.keys()
         assert rerank_result["relevance_score"] == pytest.approx(
-            invocations_result["relevance_score"], rel=0.01)
+            invocations_result["relevance_score"], rel=0.05)
```
This tolerance seems too high, we should investigate the reason behind it

maxdebayser replied:
Actually this test was a small amount over the 0.01 tolerance in another PR. Let me check if this is needed here.

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

maxdebayser committed:
address review comments

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2405,48 +2464,53 @@ def initialize_attn_backend(self, kv_cache_config: KVCacheConfig) -> None:
+        attn_layers = get_layers_from_vllm_config(self.vllm_config, Attention)
+        for attn_module in attn_layers.values():
+
+            if attn_module.attn_type == AttentionType.ENCODER_ONLY:
```
I've renamed it to attn_specs so that it doesn't imply that we're using a KV cache here. Building an array of AttentionSpec objects is not strictly necessary, but it does make it easier to reuse the code to retrieve the attention backend and the attention metadata builder.

russellb commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -2405,48 +2464,53 @@ def initialize_attn_backend(self, kv_cache_config: KVCacheConfig) -> None:
+        for attn_module in attn_layers.values():
+
+            if attn_module.attn_type == AttentionType.ENCODER_ONLY:
+                if attn_module.sliding_window is not None:
+                    attn_specs.append(
+                        SlidingWindowSpec(
+                            block_size=block_size,
+                            num_kv_heads=attn_module.num_kv_heads,
+                            head_size=attn_module.head_size,
+                            dtype=self.kv_cache_dtype,
+                            sliding_window=attn_module.sliding_window,
+                            use_mla=use_mla))
+                else:
+                    attn_specs.append(
+                        FullAttentionSpec(
+                            block_size=block_size,
+                            num_kv_heads=attn_module.num_kv_heads,
+                            head_size=attn_module.head_size,
+                            dtype=self.kv_cache_dtype,
+                            use_mla=use_mla))
```
I think sliding window may only be applicable for causal attention (DECODER), but let me know if I've misunderstood something.

```suggestion
                attn_specs.append(
                    FullAttentionSpec(
                        block_size=block_size,
                        num_kv_heads=attn_module.num_kv_heads,
                        head_size=attn_module.head_size,
                        dtype=self.kv_cache_dtype,
                        use_mla=use_mla))
```

russellb replied:
logging an error for hitting something unexpected seems OK to me

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

maxdebayser committed:
remove sliding window attention case

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
address review comment

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
make causal a flag in common attention metadata

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

russellb commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -693,6 +694,21 @@ def _prepare_inputs(
+                self.vllm_config, Attention)
+            for layer_name, attn_module in attention_layers.items():
+                if attn_module.attn_type == AttentionType.ENCODER_ONLY:
+                    attn_metadata[layer_name] = encoder_attn_metdata
```
```suggestion
                    attn_metadata[layer_name] = encoder_attn_metadata
```

maxdebayser replied:
Thanks!

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
fix typo

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

maxdebayser committed:
remove encoder model from unsupported test

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
fix apply_model tests

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

maxdebayser committed:
Try multiproc in test

To fix the test I switched to uniproc processor, but
now I'm getting weird issues like

```
torch._dynamo.exc.BackendCompilerFailed: backend='<vllm.compilation.backends.VllmBackend object at 0x7ef9f68ca600>' raised:
AttributeError: module 'torch._tensor' has no attribute 'split'
```

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
remove quant code

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

WoosukKwon reviewed (approved):
thanks for doing it! Left some comments.

maxdebayser committed:
address review comment

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into v1_encoder_only

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
trigger ci

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
@DarkLight1337 , this is approved and all tests are passing.
